tools/bridge/bridge/transfer_fetcher.py

```diff
@@ -1,25 +1,135 @@
 import logging
-
 import gevent
-from gevent.queue import Queue
 
+from gevent.queue import Queue
 from web3 import Web3
 
+MINIMAL_TOKEN_ABI = [
+    {
+        "anonymous": False,
+        "inputs": [
+            {"indexed": True, "name": "from", "type": "address"},
+            {"indexed": True, "name": "to", "type": "address"},
+            {"indexed": False, "name": "value", "type": "uint256"},
+        ],
+        "name": "Transfer",
+        "type": "event",
+    }
+]
+
+TRANSFER_EVENT_SIGNATURE_HASH = Web3.keccak(text="Transfer(address,address,uint256)")
+
+
+class TransferFetcher:
```

Comment:
A class feels a little unnecessary here, can't we just use a function? I'm imagining something like this:

```
def fetch_token_transfers(queue, w3, max_reorg_depth, transfer_fetch_limit, poll_interval):
    transfers_fetched_until = 0
    while True:
        block_number = w3.eth.blockNumber
        fetch_until = max(0, block_number - max_reorg_depth)
        if fetch_until > transfers_fetched_until:
            transfers = fetch_transfers_in_range(contract, transfers_fetched_until + 1, fetch_until)
            for transfer in transfers:
                queue.put(transfer)
    gevent.sleep(poll_interval)

def fetch_transfers_in_range(contract, start, end, transfer_fetch_limit):
    start_block_numbers = range(start + 1, end, transfer_fetch_limit)
    event_bunches = [...]
    return list(itertools.chain(event_bunches))
```

The advantage would be that it's easier to reason about, easier to test, and less likely to be abused for other stuff in the future.

Reply:
Okay, here are my thoughts why I decided to use a class:
 - reduce complexity of the individual functions to improve readability
 - having multiple functions allows for easier to handle test cases for each small "feature"
 - having a constructor function allows me to evaluate the given arguments (see the contract addresses) before a worker process becomes started

More arguments are not objective anymore. I think it is possible to rewrite all this stuff but I rather would not do it.
Your argument about abusement I do not understand. >.<

Reply:
You can split functionality up into multiple functions, a class provides no benefit here. So I don't think the first two reasons are very convincing. In fact, my argument is that having a class actually increases complexity because it makes everything stateful.

>  Your argument about abusement I do not understand.

Not a big argument, but classes tend to be extended by additional functionality whereas functions tend to stay what they originally where supposed to do. So people might add code later that doesn't fit very well but kind of works, reducing the overall quality of the code.

But if you still think a class is better, I'm ok with it, it's just a suggestion.

Reply:
If you're going to keep it: Have you considered making this a subclass of `gevent.Greenlet`? IIRC that's what you typically want to do with gevent, but @schmir might know better.

Reply:
Oh, and one more thing: We'll need the same for the confirmation event. Can we generalize this class in such a way that it can be used on any chain with any event?

Reply:
> Have you considered making this a subclass of `gevent.Greenlet`?

Just don't do that unless you have good reason to. The code as is would work with and without gevent.

Reply:
> Oh, and one more thing: We'll need the same for the confirmation event. Can we generalize this class in such a way that it can be used on any chain with any event?

yes, please do this.

Reply:
The generalization is definitely a good idea! I'm not 100% sure about every detail, but it will be worth it. :upside_down_face: 
I would argue to argue to change the name of the module then to something like `event_fetcher`?

Reply:
@jannikluhn [here](https://github.com/trustlines-protocol/blockchain/blob/feature/bridge-e-token-transfer-fetcher-showcase/tools/bridge/bridge/transfer_fetcher.py) is a version using functions only. It has 25% more lines of code. I must carry all the arguments through all functions even not everyone is using everything. Pleas confirm that you really really want to have this. 
Despite from this the generalization will happen anyways.

Reply:
I think the functions could be simplified, but by now I'm okay with the class approach (considering that we'll probably have more complicated workers that don't fit well into a function and having a shared interface for all of those is probably a good idea).

Reply:
Okay, I've generalized the class for all kind of events. Please have another look onto this part. All tests have been applied, but still test the functionality using the token and foreign bridge contract as "real" example.

Reply:
:heavy_check_mark: